sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,210 @@
+    }
+
+    private AsyncCacheNonBlocking(IEqualityComparer<TKey> keyEqualityComparer) {
+        this(new ConcurrentHashMap<>(), keyEqualityComparer);
+    }
+
+    public AsyncCacheNonBlocking() {
+        this((value1, value2) -> {
+            if (value1 == value2)
+                return true;
+            if (value1 == null || value2 == null)
+                return false;
+            return value1.equals(value2);
+        });
+    }
+
+    /**
+     *
+     * <p>
+     * If another initialization function is already running, new initialization function will not be started.
+     * The result will be result of currently running initialization function.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * Force refresh is true:
+     *  If the key does not exist: It will create and await the new task
+     *  If the key exists and the current task is still running: It will return the existing task
+     *  If the key exists and the current task is already done: It will start a new task to get the updated values.
+     *     Once the refresh task is complete it will be returned to caller.
+     *     If it is a success the value in the cache will be updated. If the refresh task throws an exception the key will be removed from the cache.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        Function<TValue, Boolean> forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("cache[{}] exists", key);
+            }
+
+            return initialLazyValue.getValueAsync().flatMap(value -> {
+                if(!forceRefresh.apply(value)) {
+                    return Mono.just(value);
+                }
+
+                Mono<TValue> refreshMono = initialLazyValue.createAndWaitForBackgroundRefreshTaskAsync(singleValueInitFunc);
+
+                return refreshMono.onErrorResume(
+                    (exception) -> {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("refresh cache [{}] resulted in error", key, exception);
+                        }
+                        if (initialLazyValue.shouldRemoveFromCache()) {
+                            this.remove(key);
```

Comment:
this seems a little bit different compared to what .net has done (like only remove for 404 exceptions)?

Reply:
Fixed the code